### PR TITLE
Migrate Customers KPIs and required fields configuration

### PR DESCRIPTION
**Pull request by sarjon**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | It migrates Customer KPIs and required fields configuration in customers list page.
| Type?         | refacto
| Category?     | BO
| BC breaks?    |no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10559
| How to test?  | Access `/admin-dev/index.php/sell/customers` page and **1.** check KPIs block, it should display same information as legacy page and **2.** "Set required fields for this section" should also work as it was working in legacy page.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11519)

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
Hello @sarjon, this needs a rebase.

File: src/Adapter/Customer/CommandHandler/SetRequiredFieldsForCustomerHandler.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * @internal
+ */
+final class SetRequiredFieldsForCustomerHandler implements SetRequiredFieldsForCustomerHandlerInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(SetRequiredFieldsForCustomerCommand $command)
+    {
+        $object = new Customer();
```
Comment (mickaelandrieu) [inline src/Adapter/Customer/CommandHandler/SetRequiredFieldsForCustomerHandler.php]: can we name it $customer ?

**Comment by sarjon** [comment]
i could rebase after review and QA validation, is that okay?

File: src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        return $this->requiredFields;
+    }
+
+    /**
+     * Check that all provided fields are allowed.
+     *
+     * @param string[] $requiredFields
+     */
+    private function assertContainsOnlyAllowedFields(array $requiredFields)
```
Comment (mickaelandrieu) [inline src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php:65]: reading this function I expect a boolean to be returned: is it intended?

File: src/Core/Domain/Customer/Exception/CannotSetRequiredFieldsForCustomerException.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Customer\Exception;
+
+/**
+ * Class CannotSetRequiredFieldsForCustomerException
```
Comment (mickaelandrieu) [inline src/Core/Domain/Customer/Exception/CannotSetRequiredFieldsForCustomerException.php]: can be improved :)

File: src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php [hunk 4]
```diff
@@ -142,4 +151,37 @@ public function viewAction($customerId, Request $request)
...
+            $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
+        }
+
+        return $this->redirectToRoute('admin_customers_index');
+    }
+
+    /**
+     * @return FormInterface
+     */
+    protected function getRequiredFieldsForm()
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php]: why this function is protected? put it private please :)

File: src/Adapter/Customer/CommandHandler/SetRequiredFieldsForCustomerHandler.php (hunk 1 above)
Comment (sarjon) [reply src/Adapter/Customer/CommandHandler/SetRequiredFieldsForCustomerHandler.php]: i guess we can, but i think the name would be wrong since its not an actual customer, but more like a service object.

File: src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php (hunk 2 above)
Comment (sarjon) [reply src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php:65]: yes, it throws exception if domain rule is violated, otherwise, everything is executed as normal.

File: src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php (hunk 4 above)
Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php]: since controller is not final, i put it as `protected`, is there are reason to make it `private`? :thinking:

File: src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        $requiredFields = (new Customer())->getFieldsRequiredDatabase();
+
+        if (empty($requiredFields)) {
+            return [];
+        }
+
+        $fields = [];
+
+        foreach ($requiredFields as $field) {
```
Comment (matks) [inline src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php:53]: `array_map()` ? Or is it less performing ?

Also please check for `field_name` array key to exist. I know that this is supposed to be ensured by database schema but if user does modify its database schema and then this page stops working, we wish he can easily find the mistake with an helpful error message 😉 . It can't hurt.

File: src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Customer\QueryHandler;
+
+use Customer;
+use PrestaShop\PrestaShop\Core\Domain\Customer\Query\GetRequiredFieldsForCustomer;
+use PrestaShop\PrestaShop\Core\Domain\Customer\QueryHandler\GetRequiredFieldsForCustomerHandlerInterface;
+
+/**
+ * Handles query which gets required fields for customer sing up
```
Comment (matks) [inline src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php]: `sing`
I am happy to know Prestashop want its customers to be singers 😃

File: src/Adapter/Kpi/AverageCustomerAgeKpi.php [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+use HelperKpi;
+use PrestaShop\PrestaShop\Core\ConfigurationInterface;
+use PrestaShop\PrestaShop\Core\Kpi\KpiInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+
+/**
+ * Renders average age of all customers.
+ */
+final class AverageCustomerAgeKpi implements KpiInterface
```
Comment (matks) [inline src/Adapter/Kpi/AverageCustomerAgeKpi.php:37]: This class can be unit tested.
Once https://github.com/PrestaShop/PrestaShop/pull/11525 is merged, we can practice heavy unit testing 🎉 !

File: src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php (hunk 4 above)
Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php]: @mickaelandrieu is against overriding controllers
On this topic I dont know yet. This is something worth discussing 🤔

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml [hunk 8]
```diff
@@ -652,3 +652,10 @@ services:
             - '@router.default'
         tags:
             - { name: form.type }
+
+    prestashop.bundle.form.form.admin.sell.customer.required_fields:
```
Comment (matks) [inline src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml]: You started using new naming convention 😛 please do it in another PR
so this PR is a "feature" type PR which means it modifies a behavior using known classes/methods
while the other PR will be a "refactoring" type PR: the app (prestashop CMS) behavior is unchanged, but the underlying implementation is modified

PR 1 : QA expects changes
PR 2: QA expects no visual changes

File: src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php (hunk 6 above)
Comment (sarjon) [reply src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php]: i always do this mistake.. :man_facepalming:

File: src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php (hunk 5 above)
Comment (sarjon) [reply src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php:53]: is it really something we have to do? i mean if user deletes table or even database, should we handle that as well? or what if we select 10 or more fields, should we check that each key exists? that way, we would add a lot of checks (which seems unnecessary) in codebase. :/

**Comment by sarjon** [comment]
comments addressed. :)

File: src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php (hunk 5 above)
Comment (matks) [reply src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php:53]: > if user deletes table or even database, should we handle that as well

IMO yes: we should handle it by throwing an exception with an helpful message that will allow user to quickly understand what is wrong.

It's 3 lines of code for us (one `if`, one `throw Exception` and one closing brack `}`) but can really help the user. Do not use a domain exception (`RuntimeException` is fine here), do not use a translated and complex error message, use some short one. So it's a small thing for us, and a great help for the PrestaShop developer.

So:
```
    if (false === array_key_exists('field_name', $field)) {
        throw new \RuntimeException('Expects property field_name in $field, but it does not exists.');
    }
    $fields[] = $field['field_name'];
}
```

Dont forget PrestaShop is a opensource CMS: people can take it and start modifying it for their needs. So they might replace the mysql database with a nosql database or an API. We will not write all our code for this usecase 😛 but the php code should check whether the input it is given is right, and crash else. It's [defensive programming](https://en.wikipedia.org/wiki/Defensive_programming).

We can discuss it with @mickaelandrieu if you want 😉 maybe I'm too paranoid but when I look at PrestaShop bugs history I think it's relevant here.

**Comment by matks** [comment]
Status: review in progress, https://github.com/PrestaShop/PrestaShop/pull/11519#pullrequestreview-178830819 to address

File: src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php (hunk 5 above)
Comment (sarjon) [reply src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php:53]: im aware of defensive programming but checking if fields exist from database seems like a big overhead. :/ 
however, i dont want this PR to be blocked by this "problem", could you create an issue in prestashop so we can discuss that as it would affect a lot of code.

Comment (matks) [reply src/Adapter/Customer/QueryHandler/GetRequiredFieldsForCustomerHandler.php:53]: All right 👍

File: src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php (hunk 2 above)
Comment (jolelievre) [reply src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php:65]: if it was named `containsOnlyAllowedFields` I would expect a boolean as well But since it starts with assert it seems logical to throw an exception
But I'm curious about where this class is caught?

File: src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        $allowedRequiredFields = [
+            RequiredField::PARTNER_OFFERS,
+        ];
+
+        if (!empty(array_diff($requiredFields, $allowedRequiredFields))) {
+            throw new InvalidCustomerRequiredFieldsException(
+                sprintf(
+                    'Invalid customer required fields provided. Allowed fields are: %s',
```
Comment (jolelievre) [inline src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php:78]: Maybe you could display which fields were refused using the diff result, to help for debugging

File: src/PrestaShopBundle/Resources/config/routing/admin/sell/customer/customers.yml [hunk 10]
```diff
@@ -38,3 +38,10 @@ admin_customers_view:
     _legacy_controller: AdminCustomers
   requirements:
     customerId: \d+
+
+admin_customers_set_required_fields:
+  path: /set-required-fields
+  methods: [POST]
+  defaults:
+    _controller: PrestaShopBundle:Admin/Sell/Customer/Customer:setRequiredFields
+    _legacy_controller: AdminCustomers
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/routing/admin/sell/customer/customers.yml:65]: could you add the `_legacy_link` as well please?

File: src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml [hunk 11]
```diff
@@ -54,3 +54,31 @@ services:
             - '@translator'
             - '@prestashop.adapter.legacy.kpi_configuration'
             - '@=service("prestashop.adapter.legacy.context").getAdminLink("AdminStats", true, {"ajax": 1, "action": "getKpi", "kpi": "products_per_category"})'
+
+    prestashop.adapter.kpi.most_common_customers_gender:
+        class: 'PrestaShop\PrestaShop\Adapter\Kpi\MostCommonCustomersGenderKpi'
+        arguments:
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml:60]: since all these classes have the same constructor maybe you could create an abstract class with a default service definition, and make the other definitions inherit it

File: src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml [hunk 12]
```diff
@@ -54,3 +54,31 @@ services:
...
+            - '@translator'
+            - '@prestashop.adapter.legacy.kpi_configuration'
+            - '@=service("prestashop.adapter.legacy.context").getAdminLink("AdminStats", true, {"ajax": 1, "action": "getKpi", "kpi": "customer_main_gender"})'
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml:63]: are you sure about these arguments, you put the arguments in `$sfRouteParams` but for legacy links you need to use `$params` so I might be wrong but I think the correct syntax would be:

```
    - '@=service("prestashop.adapter.legacy.context").getAdminLink("AdminStats", true, {}, {"ajax": 1, "action": "getKpi", "kpi": "customer_main_gender"})'
```

File: src/PrestaShopBundle/Resources/config/services/core/kpi.yml [hunk 13]
```diff
@@ -21,3 +21,11 @@ services:
           - '@prestashop.adapter.kpi.empty_categories'
           - '@prestashop.adapter.kpi.top_category'
           - '@prestashop.adapter.kpi.average_products_in_category'
+
+    prestashop.core.kpi_row.factory.customers:
+        class: PrestaShop\PrestaShop\Core\Kpi\Row\KpiRowFactory
+        arguments:
+            - '@prestashop.adapter.kpi.most_common_customers_gender'
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/core/kpi.yml:28]: I'm wondering if those adapter services should be in the core
maybe it would me more correct to declare aliases like:

```
    prestashop.core.kpi.most_common_customers_gender: '@prestashop.adapter.kpi.most_common_customers_gender'
```

so that core services fiels only point to core services, and we will only have to redeclare the proper service in the core later?

(not necessarily relevant for this specific PR)

**Review by jolelievre — COMMENTED** [review]
nice job @sarjon 👍

File: src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php (hunk 9 above)
Comment (sarjon) [reply src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php:78]: i think adding allowed fields gives developer the idea what is wrong. Adding diff would provide same information just from other way around, no? it the same as with `OptionsResolver` provided by symfony, it only dump the supported keys?

File: src/PrestaShopBundle/Resources/config/routing/admin/sell/customer/customers.yml (hunk 10 above)
Comment (sarjon) [reply src/PrestaShopBundle/Resources/config/routing/admin/sell/customer/customers.yml:65]: ill add it when we enable whole page. :+1:

File: src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml (hunk 11 above)
Comment (sarjon) [reply src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml:60]: maybe, but in another PR, there are quite more KPI classes.

File: src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml (hunk 12 above)
Comment (sarjon) [reply src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml:63]: actually, current argument order works well, are you sure it wouldnt break? :thinking:

File: src/PrestaShopBundle/Resources/config/services/core/kpi.yml (hunk 13 above)
Comment (sarjon) [reply src/PrestaShopBundle/Resources/config/services/core/kpi.yml:28]: i dont know really.. :confused:  @matks wdyt?

**Comment by sarjon** [comment]
@matks bring back `waiting for QA`. :smiling_imp:

**Comment by matks** [comment]
I removed it so that if you modified the PR because of jolelievre's feedback, QA would validate the "latest" version 😉 
Let's go then !

**Review by matks — DISMISSED** [review]

**Comment by marionf** [comment]
Hello @sarjon 

The average age isn't good, it takes the same value as Order per customer

capture d ecran_694

capture d ecran_693


About required fields, the information message is missing

capture d ecran_695

capture d ecran_696

It should looks like this:

capture d ecran_697

**Comment by sarjon** [comment]
@marionf both issues are fixed. :)

File: src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php (hunk 9 above)
Comment (jolelievre) [reply src/Core/Domain/Customer/Command/SetRequiredFieldsForCustomerCommand.php:78]: I didn't mean not to display allowed fields, but also show which field is incorrect Quite often you just mistyped a field and displaying the wrong next to the correct one helps you see the mistake.

File: src/PrestaShopBundle/Resources/config/routing/admin/sell/customer/customers.yml (hunk 10 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/routing/admin/sell/customer/customers.yml:65]: ok great!

File: src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml (hunk 12 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/services/adapter/kpi.yml:63]: oh right, I understood why
That's because the method `getAdminLink` from `LegacyContext` uses the same argument for both third and forth argument in `LinkCore::getAdminLink`
My bad!

**Comment by sarjon** [comment]
@matks this one is candidate for merge. ;)

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
Thanks @sarjon 😄
